internal/v2/cert_verifier/cert_verifier.go

```diff
@@ -3,64 +3,83 @@ package certverifier
 
 import (
 	"crypto/x509"
-	"fmt"
-	"time"
+
+	s2av2pb "github.com/google/s2a-go/internal/proto/v2/s2a_go_proto"
 )
 
 // VerifyClientCertificateChain verifies a client's certificate chain against
 // the roots in the provided pool and checks that the common name in the
 // client's leaf certificate matches the expected common name.
-func VerifyClientCertificateChain(expectedCommonName string, pool *x509.CertPool) func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+func VerifyClientCertificateChain(pool *x509.CertPool, cstream s2av2pb.S2AService_SetUpSessionClient) func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
 	return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
-		if len(rawCerts) == 0 {
-			return nil
-		}
-		cert, err := x509.ParseCertificate(rawCerts[0])
-		if err != nil {
-			return fmt.Errorf("ParseCertificate failed: %v", err)
-		}
-
-		opts := x509.VerifyOptions{
-			CurrentTime: time.Now(),
-			Roots:       pool,
+		// Offload verification to S2Av2.
+		if err := cstream.Send(&s2av2pb.SessionReq {
+			AuthenticationMechanisms: []*s2av2pb.AuthenticationMechanism {
+				{
+					// TODO(rmehta19): Populate Authentication Mechanism using tokenmanager.
+					MechanismOneof: &s2av2pb.AuthenticationMechanism_Token{"token"},
+				},
+			},
+			ReqOneof: &s2av2pb.SessionReq_ValidatePeerCertificateChainReq {
+				&s2av2pb.ValidatePeerCertificateChainReq {
+					Mode: s2av2pb.ValidatePeerCertificateChainReq_SPIFFE,
+					PeerOneof: &s2av2pb.ValidatePeerCertificateChainReq_ClientPeer_ {
+						&s2av2pb.ValidatePeerCertificateChainReq_ClientPeer {
+							CertificateChain: rawCerts,
```

Comment:
Can you confirm that the order of certs in rawCerts is what CertificateChain is expecting?

Reply:
Do you mean add a function to verify rawCerts is indeed a list of DER encoded certs ordered from leaf to root?

CertificateChain expects the DER encoded certs to be ordered from leaf to root:
https://github.com/google/s2a-go/blob/23127301b4c006bbb890458e74ce23f8d64cf719/internal/proto/v2/s2a_go_proto/s2a.pb.go#L1579

Reply:
Discussed offline: go will always pass from leaf to root, and in unit tests I also pass from leaf to root. To clear up things I will
- rename the example cert files with more descriptive names
- create a README explaining how I generate the example certs
- add a test case where I pass certs from root to leaf and verify an error occurs
- add a test case testing the intermediates field of Verify

I'll probably not do all these things in this PR, but in follow up PR's.

Reply:
SGTM. Thanks Riya!